lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+
+            {
+              {lower_bound, params},
+              {iter + 1, converged, x, key, max_iter, tol, reg_covar}
+            }
+          end
+
+        {max_lower_bound, best_params} =
+          if max_lower_bound < lower_bound do
+            {lower_bound, params}
+          else
+            {max_lower_bound, best_params}
+          end
+
+        {
+          best_params,
+          {run + 1, max_lower_bound, x, key, max_iter, tol, reg_covar}
+        }
+      end
+
+    {weights, means, covariances, precisions_cholesky} = best_params
+
+    %__MODULE__{
+      weights: weights,
+      means: means,
+      covariances: covariances,
+      precisions_cholesky: precisions_cholesky
+    }
+  end
+
+  defnp init_params(key, x, opts) do
+    type = Nx.type(x)
+    {num_samples, _} = Nx.shape(x)
+    num_gaussians = opts[:num_gaussians]
+    reg_covar = opts[:reg_covar]
+
+    k_means_model =
+      Scholar.Cluster.KMeans.fit(x, num_clusters: num_gaussians, num_runs: 1, key: key)
+
+    labels = k_means_model.labels
+    resp = Nx.take(Nx.eye(num_gaussians, type: type), labels)
+
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, resp, reg_covar)
+
+    weights = weights / num_samples
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_resp} = estimate_log_prob_resp(x, weights, means, precisions_cholesky)
+    {Nx.mean(log_prob_norm), log_resp}
+  end
+
+  defnp m_step(x, log_resp, reg_covar) do
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, Nx.exp(log_resp), reg_covar)
+
+    weights = weights / Nx.sum(weights)
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_parameters(x, resp, reg_covar) do
+    nk = Nx.sum(resp, axes: [0])
+    means = Nx.dot(Nx.transpose(resp), x) / nk
+    covariances = estimate_gaussian_covariances(x, resp, nk, means, reg_covar)
+    precisions_cholesky = compute_precisions_cholesky(covariances)
+    {nk, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_covariances(x, resp, nk, means, reg_covar) do
+    type = Nx.type(x)
+    {num_gaussians, num_features} = Nx.shape(means)
+
+    {covariances, _} =
+      while {
+              covariances =
+                Nx.tensor(0.0, type: type)
+                |> Nx.broadcast({num_gaussians, num_features, num_features}),
+              {k = 0, x, resp, nk, means, reg_covar}
+            },
+            k < num_gaussians do
+        diff = x - means[k]
+        covariance = Nx.dot(resp[[.., k]] * Nx.transpose(diff), diff) / nk[k]
+        covariance = Nx.put_diagonal(covariance, Nx.take_diagonal(covariance) + reg_covar)
+
+        {
+          Nx.put_slice(covariances, [k, 0, 0], Nx.new_axis(covariance, 0)),
+          {k + 1, x, resp, nk, means, reg_covar}
+        }
+      end
```

Comment:
This while also reads like something that can benefit from vectorization

Reply:
Also, we can leave usage of vectorization as TO-DO comments instead, unless you really want to tackle this. It might even be the case where we merge this and you tackle it afterwards. Thoughts?

Reply:
Indeed, this might be vectorizable. I would need to think about lines 255 and 256 though as it's not that straightforward.
I would say let's merge it like this and I can continue working on it (I intend to continue contributing to `Nx` anyway).

Reply:
Sure! To be clear, vectorization is not a blocker at all for this PR